sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala

```diff
@@ -2065,29 +2065,238 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
-            StructType(StructField("f", StringType) :: StructField("f0", LongType) :: Nil)
-          checkAnswer(df, Seq(("a", 1), ("c", 2), ("d", 3)).toDF())
-          assert(df.schema === expectedSchema)
+  def testFile(fileName: String): String = {
+    Thread.currentThread().getContextClassLoader.getResource(fileName).toString
+  }
+
+  test("SPARK-23723: json in UTF-16 with BOM") {
+    val fileName = "json-tests/utf16WithBOM.json"
+    val schema = new StructType().add("firstName", StringType).add("lastName", StringType)
+    val jsonDF = spark.read.schema(schema)
+      // This option will be replaced by .option("lineSep", "x00 0a")
+      // as soon as lineSep allows to specify sequence of bytes in hexadecimal format.
+      .option("mode", "DROPMALFORMED")
+      .json(testFile(fileName))
+
+    checkAnswer(jsonDF, Seq(
+      Row("Chris", "Baird"), Row("Doug", "Rood")
+    ))
+  }
+
+  test("SPARK-23723: multi-line json in UTF-32BE with BOM") {
+    val fileName = "json-tests/utf32BEWithBOM.json"
+    val schema = new StructType().add("firstName", StringType).add("lastName", StringType)
+    val jsonDF = spark.read.schema(schema)
+      .option("multiline", "true")
+      .json(testFile(fileName))
+
+    checkAnswer(jsonDF, Seq(Row("Chris", "Baird")))
+  }
+
+  test("SPARK-23723: Use user's encoding in reading of multi-line json in UTF-16LE") {
+    val fileName = "json-tests/utf16LE.json"
+    val schema = new StructType().add("firstName", StringType).add("lastName", StringType)
+    val jsonDF = spark.read.schema(schema)
+      .option("multiline", "true")
+      .options(Map("encoding" -> "UTF-16LE"))
+      .json(testFile(fileName))
+
+    checkAnswer(jsonDF, Seq(Row("Chris", "Baird")))
+  }
+
+  test("SPARK-23723: Unsupported charset name") {
+    val invalidCharset = "UTF-128"
+    val exception = intercept[java.io.UnsupportedEncodingException] {
+      spark.read
+        .options(Map("charset" -> invalidCharset, "lineSep" -> "\n"))
+        .json(testFile("json-tests/utf16LE.json"))
+        .count()
+    }
+
+    assert(exception.getMessage.contains(invalidCharset))
+  }
+
+  test("SPARK-23723: checking that the charset option is case agnostic") {
+    val fileName = "json-tests/utf16LE.json"
+    val schema = new StructType().add("firstName", StringType).add("lastName", StringType)
+    val jsonDF = spark.read.schema(schema)
+      .option("multiline", "true")
+      .options(Map("charset" -> "uTf-16lE"))
+      .json(testFile(fileName))
+
+    checkAnswer(jsonDF, Seq(Row("Chris", "Baird")))
+  }
+
+
+  test("SPARK-23723: specified charset is not matched to actual charset") {
+    val fileName = "json-tests/utf16LE.json"
+    val schema = new StructType().add("firstName", StringType).add("lastName", StringType)
+    val exception = intercept[SparkException] {
+      spark.read.schema(schema)
+        .option("mode", "FAILFAST")
+        .option("multiline", "true")
+        .options(Map("charset" -> "UTF-16BE"))
```

Comment:
Can we have a test case with `UTF-16`? I assume the delimiter itself has the BOM and won't work correctly?

Reply:
If that's difficult (or weird or hacky), let's just make whitelist and document them explicitly.

Reply:
You don't have to document 100% list. just best effort thing. We can fix and add more encodings later if anyone found later, if it's hard to find out every encoding supported.

Reply:
I would blacklist 2 encodings for now - UTF-16 and UTF-32 but allow UTF-8, UTF-16LE/BE, UTF-32LE/BE and others like ISO 8859-1.

Reply:
yup that'd work too.